planning_outlines/spectralcube-to-specutils/plan.md

```diff
@@ -0,0 +1,61 @@
+* `spectral_slab` - note however that this is another word for the existing spectral region machinery in `specutils.manipulation.extract_region`, just with slightly different syntax. So `spectral_slab` might just be an alternate name that would be adopted in a `RadioSpectrum1D` or similar.
+* Moment maps: Not this is already possible in `specutils`, particularly for first- (np.mean) and second- (line width) maps, but exposing a generic "nth moment" function is a good idea.
+* `minimal_subcube` (see spectral-cube docs), although this would need a different name in a `Spectrum1D` context (since it would also potentially mask off the spectral "edges"?)
+* Spectral extraction (based on astropy regions/photutils?)
+* `with_mask` convenience method (not mandatory, but spectral-cube demonstrates it's use as syntactic sugar?)
+* Lazy masking (may not be a requirement?)
+
+
+# Problems/incompatibilities and solutions
+
+## Problem: the name `Spectrum1D`/`SpectralCollection`
+
+A user may not naturally think "cube" when they see "1d" or even "collection".
+
+### Solution
+
+1. Documentation/explanation that yes, a cube can be "1D" in the sense that it's rectified.  Caveat: people don't read documentation!
+2. Set up some aliases or subclasses with specific conventions.  E.g. `specutils.RectifiedSpectralCube` could be a subclass of `Spectrum1D` but just with an initializer that checks that the wcs has a spatial and spectral component.
+3. Point users to `NDCube` as a way to see explicit cube language, but have them use the NDCube<->Spectrum1D compatibility layer as the way to still use `specutils` functionality
+
+## Problem: large cubes/those that don't fit in memory
+
+`SpectralCube` takes some care to not copy cubes when possible.  It also provides hints on how to use memory mapping (from fits files) to handle cubes that are too big to fit in memory.  It is not clear if `specutils` is as careful about this or not.
+
+### Solution
+
+Run side-by-side tests of `spectral-cube` operations and the equivalent `specutils` operations and see if the later uses more memory.  Where this occurs, identify where the copying occurs and fix it.  Note this is a potentially unbounded amount of work, so some careful prioritization may be needed.
+
+## Incompatibility: masking convention
+
+The `spectral-cube` masking convention is the opposite of that used in specutils (which is the one used in numpy), and instead matches "indexing" convention.  This means code written for `spectral-cube` is subtle different.
+
+### Solution
+
+1. Don't fix.  Just accept that there's not direct-API compatibility
+2. Add a spectral-cube-like compatibility layer that auto inverts the masks - this isn't too hard because masking in `spectral-cube` is done via a `with_mask` method instead of direct indexing.  But it might make for more confusion because then there's two ways to do the same thing.
+
+
```

Comment:
Similar issue: indexing convention for JWST cube dimensions is different, and JWST cubes need to be transposed to be used by Spectrum1D.

Reply:
Ah good point, will add that too.

Reply:
Although after some thought I think this isn't quite as bad because it can be fixed at the loader level (as opposed to `spectral-cube` which is an API so we can't post-fix it).  Either way, added some words to the notebook about it.